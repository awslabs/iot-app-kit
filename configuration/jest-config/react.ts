export default {
  setupFilesAfterEnv: ['jest-extended/all'],
  maxWorkers: '50%',
  moduleFileExtensions: ['js', 'ts', 'jsx', 'tsx'],
  collectCoverage: true,
  testEnvironment: 'jsdom',
  transform: {
    '.*\\.(tsx?|jsx?)$': [
      '@swc/jest',
      {
        jsc: {
          transform: {
            react: {
              runtime: 'automatic',
            },
          },
        },
      },
    ],
  },
  transformIgnorePatterns: [],
  moduleNameMapper: {
    '\\.(css|scss)$': 'identity-obj-proxy',
  },
};
