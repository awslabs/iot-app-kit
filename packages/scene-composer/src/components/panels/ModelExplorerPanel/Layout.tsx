import React, { FC } from 'react';
import { FormattedMessage, defineMessages } from 'react-intl';

import LogProvider from '../../../logger/react-logger/log-provider';

const strings = defineMessages({
  NoModelSelected: {
    defaultMessage: 'Select a model ref node to explore the subtree',
    description:
      'A message displayed in a panel to indicate to the user they need to select an object in the scene in order for this panel to be populated.',
  },
});

interface ModelExplorerLayoutProps {
  active: boolean;
}

const ModelExplorerLayout: FC<ModelExplorerLayoutProps> = ({ active, children }) => {
  return (
    <LogProvider namespace='ModelExplorer'>
      {active && children}
      {!active && <FormattedMessage tagName={'p'} {...strings.NoModelSelected} />}
    </LogProvider>
  );
};

export default ModelExplorerLayout;
