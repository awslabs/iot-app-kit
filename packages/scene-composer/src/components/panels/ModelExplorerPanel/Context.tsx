import { createContext, useContext } from 'react';
import { Object3D, Event } from 'three';

interface ModelExplorerState {
  object3D?: Object3D<Event>;
}

const ModelExplorerContext = createContext<ModelExplorerState>({});

export const useModelExplorer = () => {
  return useContext(ModelExplorerContext);
};

export default ModelExplorerContext;
