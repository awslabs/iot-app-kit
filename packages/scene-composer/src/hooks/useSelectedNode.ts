import { useCallback, useMemo } from 'react';

import { useSceneComposerId } from '../common/sceneComposerIdContext';
import { useEditorState, useStore } from '../store';

const useSelectedNode = () => {
  const sceneComposerId = useSceneComposerId();
  const {
    selectedSceneNodeRef,
    selectedSceneSubmodelRef,
    getObject3DBySceneNodeRef,
    setSelectedSceneNodeRef,
    setSelectedSceneSubmodelRef,
  } = useEditorState(sceneComposerId);

  const getSceneNodeByRef = useStore(sceneComposerId)((state) => state.getSceneNodeByRef);

  const selectedSceneNode = useMemo(() => getSceneNodeByRef(selectedSceneNodeRef), [selectedSceneNodeRef]);

  // callback, not memo because this object could be very large, and we want to avoid creating multiple copies in memory
  const getSelectedObject = useCallback(() => {
    return getObject3DBySceneNodeRef(selectedSceneNodeRef);
  }, [selectedSceneNodeRef, getObject3DBySceneNodeRef]);

  const getSelectedSubmodel = useCallback(() => {
    const object3D = getSelectedObject();

    if (object3D) {
      if (typeof selectedSceneSubmodelRef === 'number') {
        return object3D.getObjectById(Number(selectedSceneSubmodelRef));
      }

      return object3D.getObjectByName(selectedSceneSubmodelRef as string);
    }
  }, [selectedSceneSubmodelRef]);

  return {
    selectedSceneNodeRef,
    selectedSceneNode,
    setSelectedSceneNodeRef,
    selectedSceneSubmodelRef,
    setSelectedSceneSubmodelRef,
    getSelectedObject,
    getSelectedSubmodel,
  };
};

export default useSelectedNode;
