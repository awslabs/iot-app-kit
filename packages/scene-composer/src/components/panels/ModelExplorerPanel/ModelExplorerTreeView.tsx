import React, { FC, useCallback, useEffect, useMemo, useState } from 'react';
import { Object3D, Event, Color, Mesh } from 'three';

import useMaterialEffect from '../../../hooks/useMaterialEffect';
import useSelectedNode from '../../../hooks/useSelectedNode';
import Tree, { TreeItem } from '../../Tree';
import { ISubModelRefComponent, KnownComponentType } from '../../../interfaces';
import { ISceneComponentInternal, ISceneNodeInternal, useStore } from '../../../store';
import { useSceneComposerId } from '../../../common/sceneComposerIdContext';
import { generateUUID } from '../../../utils/mathUtils';

import { useModelExplorer } from './Context';

export interface ModelExplorerTreeProps {
  object: Object3D<Event>;
  selected?: boolean;
  selectable?: boolean;
  expanded?: boolean;
}

interface TreeItemLabelProps {
  onCreateRef: () => void;
}

const TreeItemLabel: FC<TreeItemLabelProps> = ({ onCreateRef, children }) => {
  return <span onDoubleClick={onCreateRef}>{children}</span>;
};

export const ModelExplorerTree: FC<ModelExplorerTreeProps> = ({
  object,
  expanded: defaultExpanded = true,
  selectable = true,
}) => {
  const sceneComposerId = useSceneComposerId();
  const appendSceneNode = useStore(sceneComposerId)((state) => state.appendSceneNode);
  const setSceneNodeObject3DMapping = useStore(sceneComposerId)((state) => state.setSceneNodeObject3DMapping);
  const [expanded, setExpanded] = useState(defaultExpanded);

  const { setSelectedSceneSubmodelRef, getSelectedSubmodel, selectedSceneNodeRef } = useSelectedNode();

  const selected = useMemo(() => object === getSelectedSubmodel(), [getSelectedSubmodel]);

  const selectedColor = new Color(0x00ff00);

  const { name, children: allNodes } = object;
  const nodes = allNodes.filter((n) => !!n.name);

  const [transform, restore] = useMaterialEffect((o) => {
    if (o instanceof Mesh) {
      o.material.color = selectedColor;
    }
  }, object);

  useEffect(() => {
    if (selected) {
      transform();
    } else {
      restore();
    }
  }, [selected]);

  const onSelected = useCallback((s) => {
    setSelectedSceneSubmodelRef(s ? object.id || object.name : undefined);
  }, []);

  const onCreate = useCallback(() => {
    const nodeRef = `${selectedSceneNodeRef}#${object.id}`;
    const subModelComponent: ISubModelRefComponent = {
      type: KnownComponentType.SubModelRef,
      parentRef: selectedSceneNodeRef,
      ref: generateUUID(),
      selector: object.name,
    };

    const node = {
      ref: nodeRef,
      name: object.name,
      components: [subModelComponent as ISceneComponentInternal],
      parentRef: selectedSceneNodeRef,
    } as ISceneNodeInternal;

    appendSceneNode(node);
    setSceneNodeObject3DMapping(nodeRef, object); // Cache Reference
  }, [object]);

  return (
    <TreeItem
      labelText={<TreeItemLabel onCreateRef={onCreate}>{name}</TreeItemLabel>}
      expandable={nodes.length > 0}
      expanded={expanded}
      onExpand={setExpanded}
      onSelected={onSelected}
      selected={selected}
      selectable={selectable}
    >
      {nodes.length > 0 && (
        <Tree>
          {nodes.map((c) => (
            <ModelExplorerTree key={c.id} object={c} selectable={selectable && !selected} />
          ))}
        </Tree>
      )}
    </TreeItem>
  );
};

const ModelExplorerTreeView = () => {
  const { object3D } = useModelExplorer();

  return <Tree>{object3D && <ModelExplorerTree object={object3D} />}</Tree>;
};

export default ModelExplorerTreeView;
