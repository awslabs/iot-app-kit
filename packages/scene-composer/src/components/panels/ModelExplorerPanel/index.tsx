import React from 'react';

import useSelectedNode from '../../../hooks/useSelectedNode';

import ModelExplorerLayout from './Layout';
import ModelExplorerContext from './Context';
import ModelExplorerTreeView from './ModelExplorerTreeView';

const ModelExplorerPanel = () => {
  const { getSelectedObject } = useSelectedNode();
  const scene = getSelectedObject();

  // istanbul ignore next: nullish
  const object3D = scene?.getObjectByName('Scene') || scene;

  return (
    <ModelExplorerLayout active={!!scene}>
      <ModelExplorerContext.Provider
        value={{
          object3D,
        }}
      >
        <ModelExplorerTreeView />
      </ModelExplorerContext.Provider>
    </ModelExplorerLayout>
  );
};

export default ModelExplorerPanel;
