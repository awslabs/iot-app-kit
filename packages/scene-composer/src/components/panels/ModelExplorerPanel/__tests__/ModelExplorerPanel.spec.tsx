import React from 'react';
import { render } from '@testing-library/react';

import useSelectedNode from '../../../../hooks/useSelectedNode';
import ModelExplorerPanel from '..';

jest.mock('../Layout', () => (props) => <div data-mocked='Layout' {...props} />);
jest.mock('../ModelExplorerTreeView', () => (props) => <div data-mocked='ModelExplorerTreeView' {...props} />);
jest.mock('../../../../hooks/useSelectedNode');

describe('<ModelExplorerPanel />', () => {
  it('should render as expected', () => {
    const getObjectByName = jest.fn(() => ({ name: 'Scene', children: [] }));

    const scene = {
      name: 'Root',
      children: [
        {
          name: 'Bcene',
          children: [
            { name: 'Scene', children: [] },
            { name: 'Clean', children: [] },
          ],
        },
      ],
      getObjectByName,
    };

    (useSelectedNode as jest.Mock).mockImplementation(() => ({ getSelectedObject: () => scene }));

    const { container } = render(<ModelExplorerPanel />);

    expect(container).toMatchSnapshot();
    expect(getObjectByName).toBeCalledWith('Scene');
  });
});
