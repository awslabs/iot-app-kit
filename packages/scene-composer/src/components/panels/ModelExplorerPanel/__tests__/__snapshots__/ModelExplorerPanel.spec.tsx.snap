// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModelExplorerPanel /> should render as expected 1`] = `
<div>
  <div
    data-mocked="Layout"
  >
    <div
      data-mocked="ModelExplorerTreeView"
    />
  </div>
</div>
`;
